File: app/src/main/java/org/breezyweather/common/ui/widgets/horizontal/HorizontalRecyclerView.java
package org.breezyweather.common.ui.widgets.horizontal;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.ViewConfiguration;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

public class HorizontalRecyclerView extends RecyclerView {

    private int mPointerId;
    private float mInitialX;
    private float mInitialY;
    private final int mTouchSlop;
    private boolean mBeingDragged;
    private boolean mHorizontalDragged;

    private static final String TAG = "HorizontalRecyclerView";

    public HorizontalRecyclerView(Context context) {
        this(context, null);
    }

    public HorizontalRecyclerView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public HorizontalRecyclerView(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        mTouchSlop = ViewConfiguration.get(getContext()).getScaledTouchSlop();
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent ev) {
        switch (ev.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                mBeingDragged = false;
                mHorizontalDragged = false;

                mPointerId = ev.getPointerId(0);
                mInitialX = ev.getX();
                mInitialY = ev.getY();

                getParent().requestDisallowInterceptTouchEvent(true);
                break;

            case MotionEvent.ACTION_POINTER_DOWN: {
                int index = ev.getActionIndex();
                mPointerId = ev.getPointerId(index);
                mInitialX = ev.getX(index);
                mInitialY = ev.getY(index);
                break;
            }
            case MotionEvent.ACTION_MOVE: {
                int index = ev.findPointerIndex(mPointerId);
                if (index == -1) {
                    Log.e(TAG, "Invalid pointerId=" + mPointerId + " in onTouchEvent");
                    break;
                }

                float x = ev.getX(index);
                float y = ev.getY(index);

                if (!mBeingDragged && !mHorizontalDragged) {
                    if (Math.abs(x - mInitialX) > mTouchSlop || Math.abs(y - mInitialY) > mTouchSlop) {
                        mBeingDragged = true;
                        if (Math.abs(x - mInitialX) > Math.abs(y - mInitialY)) {
                            mHorizontalDragged = true;
                        } else {
                            getParent().requestDisallowInterceptTouchEvent(false);
                        }
                    }
                }
                break;
            }
            case MotionEvent.ACTION_POINTER_UP: {
                int index = ev.getActionIndex();
                int id = ev.getPointerId(index);
                if (mPointerId == id) {
                    int newIndex = index == 0 ? 1 : 0;

                    this.mPointerId = ev.getPointerId(newIndex);
                    mInitialX = (int) ev.getX(newIndex);
                    mInitialY = (int) ev.getY(newIndex);
                }
                break;
            }
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                mBeingDragged = false;
                mHorizontalDragged = false;
                getParent().requestDisallowInterceptTouchEvent(false);
                break;
        }

        return super.onInterceptTouchEvent(ev) && mBeingDragged && mHorizontalDragged;
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        switch (ev.getActionMasked()) {
            case MotionEvent.ACTION_POINTER_DOWN: {
                int index = ev.getActionIndex();
                mPointerId = ev.getPointerId(index);
                mInitialX = ev.getX(index);
                mInitialY = ev.getY(index);
                break;
            }
            case MotionEvent.ACTION_MOVE: {
                int index = ev.findPointerIndex(mPointerId);
                if (index == -1) {
                    Log.e(TAG, "Invalid pointerId=" + mPointerId + " in onTouchEvent");
                    break;
                }

                float x = ev.getX(index);
                float y = ev.getY(index);

                if (!mBeingDragged && !mHorizontalDragged) {
                    mBeingDragged = true;
                    if (Math.abs(x - mInitialX) > Math.abs(y - mInitialY)) {
                        mHorizontalDragged = true;
                    } else {
                        getParent().requestDisallowInterceptTouchEvent(false);
                    }
                }
                break;
            }
            case MotionEvent.ACTION_POINTER_UP: {
                int index = ev.getActionIndex();
                int id = ev.getPointerId(index);
                if (mPointerId == id) {
                    int newIndex = index == 0 ? 1 : 0;

                    this.mPointerId = ev.getPointerId(newIndex);
                    mInitialX = (int) ev.getX(newIndex);
                    mInitialY = (int) ev.getY(newIndex);
                }
                break;
            }
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                mBeingDragged = false;
                mHorizontalDragged = false;
                getParent().requestDisallowInterceptTouchEvent(false);
                break;
        }

        return super.onTouchEvent(ev);
    }
}

File: app/src/main/java/org/breezyweather/ui/main/widgets/HorizontalRecyclerView.kt
/*
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */
package org.breezyweather.ui.main.widgets

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.ViewConfiguration
import androidx.recyclerview.widget.RecyclerView
import kotlin.math.abs

/**
 * This recycler view handles touch events when used inside an already horizontally scrollable/swipeable container
 */
open class HorizontalRecyclerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0,
) : RecyclerView(context, attrs, defStyle) {
    private var mPointerId = 0
    private var mInitialX = 0f
    private var mInitialY = 0f
    private val mTouchSlop = ViewConfiguration.get(getContext()).scaledTouchSlop
    private var mBeingDragged = false
    private var mHorizontalDragged = false

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        when (ev.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                mBeingDragged = false
                mHorizontalDragged = false

                mPointerId = ev.getPointerId(0)
                mInitialX = ev.x
                mInitialY = ev.y

                parent.requestDisallowInterceptTouchEvent(true)
            }

            MotionEvent.ACTION_POINTER_DOWN -> {
                val index = ev.actionIndex
                mPointerId = ev.getPointerId(index)
                mInitialX = ev.getX(index)
                mInitialY = ev.getY(index)
            }

            MotionEvent.ACTION_MOVE -> {
                val index = ev.findPointerIndex(mPointerId)
                if (index == -1) {
                    Log.e(TAG, "Invalid pointerId=$mPointerId in onTouchEvent")
                } else {
                    val x = ev.getX(index)
                    val y = ev.getY(index)

                    if (!mBeingDragged && !mHorizontalDragged) {
                        if (abs(x - mInitialX) > mTouchSlop || abs(y - mInitialY) > mTouchSlop) {
                            mBeingDragged = true
                            if (abs(x - mInitialX) > abs(y - mInitialY)) {
                                mHorizontalDragged = true
                            } else {
                                parent.requestDisallowInterceptTouchEvent(false)
                            }
                        }
                    }
                }
            }

            MotionEvent.ACTION_POINTER_UP -> {
                val index = ev.actionIndex
                val id = ev.getPointerId(index)
                if (mPointerId == id) {
                    val newIndex = if (index == 0) 1 else 0

                    this.mPointerId = ev.getPointerId(newIndex)
                    mInitialX = ev.getX(newIndex).toInt().toFloat()
                    mInitialY = ev.getY(newIndex).toInt().toFloat()
                }
            }

            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                mBeingDragged = false
                mHorizontalDragged = false
                parent.requestDisallowInterceptTouchEvent(false)
            }
        }

        return super.onInterceptTouchEvent(ev) && mBeingDragged && mHorizontalDragged
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(ev: MotionEvent): Boolean {
        when (ev.actionMasked) {
            MotionEvent.ACTION_POINTER_DOWN -> {
                val index = ev.actionIndex
                mPointerId = ev.getPointerId(index)
                mInitialX = ev.getX(index)
                mInitialY = ev.getY(index)
            }

            MotionEvent.ACTION_MOVE -> {
                val index = ev.findPointerIndex(mPointerId)
                if (index == -1) {
                    Log.e(TAG, "Invalid pointerId=$mPointerId in onTouchEvent")
                } else {
                    val x = ev.getX(index)
                    val y = ev.getY(index)

                    if (!mBeingDragged && !mHorizontalDragged) {
                        mBeingDragged = true
                        if (abs(x - mInitialX) > abs(y - mInitialY)) {
                            mHorizontalDragged = true
                        } else {
                            parent.requestDisallowInterceptTouchEvent(false)
                        }
                    }
                }
            }

            MotionEvent.ACTION_POINTER_UP -> {
                val index = ev.actionIndex
                val id = ev.getPointerId(index)
                if (mPointerId == id) {
                    val newIndex = if (index == 0) 1 else 0

                    this.mPointerId = ev.getPointerId(newIndex)
                    mInitialX = ev.getX(newIndex).toInt().toFloat()
                    mInitialY = ev.getY(newIndex).toInt().toFloat()
                }
            }

            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                mBeingDragged = false
                mHorizontalDragged = false
                parent.requestDisallowInterceptTouchEvent(false)
            }
        }

        return super.onTouchEvent(ev)
    }

    companion object {
        private const val TAG = "HorizontalRecyclerView"
    }
}
